tests/unit/Utils/DPOPTest.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+namespace Pdsinterop\Solid\Auth\Utils;
+
+use PHPUnit\Framework\TestCase;
+use Lcobucci\JWT\Validation\RequiredConstraintsViolated;
+
+/**
+ * @covers DPop
+ */
+class DPOPTest extends TestCase
+{
+
+	private $dpop;
+	private $url;
+	private $serverRequest;
+
+	protected function sign($dpop, $privateKey=null)
+	{
+		$keyPath    = dirname(__DIR__) . '/../fixtures/keys';
+		if (!$privateKey) {
+	        $privateKey = file_get_contents($keyPath . '/private.key');
+	    }
+
+		$signature = '';
+        $success = \openssl_sign(
+            Base64Url::encode(json_encode($dpop['header'])).'.'.
+            Base64Url::encode(json_encode($dpop['payload'])), 
+            $signature, 
+            $privateKey, 
+            OPENSSL_ALGO_SHA256
+        );
+
+        if (!$success) {
+        	throw new \Exception('unable to sign dpop');
+        }
+        $token = Base64Url::encode(json_encode($dpop['header'])).'.'.
+            	 Base64Url::encode(json_encode($dpop['payload'])).'.'.
+            	 Base64Url::encode($signature);
+
+        return array_merge($dpop, [
+        	'signature' => $signature,
+        	'token' => $token
+        ]);
+	}
+
+	protected function setUp(): void
+	{
+		$keyPath    = dirname(__DIR__) . '/../fixtures/keys';
+        $privateKey = file_get_contents($keyPath . '/private.key');
+        $publicKey  = file_get_contents($keyPath . '/public.key');
+
+        $keyInfo = \openssl_pkey_get_details(\openssl_pkey_get_public($publicKey));
+        $jwk = [
+            'kty' => 'RSA',
+            'n' => Base64Url::encode($keyInfo['rsa']['n']),
+            'e' => Base64Url::encode($keyInfo['rsa']['e'])
+        ];
+
+        $header = [
+            'typ' => 'dpop+jwt',
+            'alg' => 'RS256',
+            'jwk' => $jwk
+        ];
+
+        $payload = [
+            'iss' => 'example.com',
+            'aud' => 'example.com',
+            'htm' => 'GET',
+            'htu' => 'https://www.example.com',
+            'iat' => time(),
+            'nbf' => time(),
+            'exp' => time()+3600
+        ];
+
+        $this->dpop = $this->sign([
+        	'header'  => $header,
+        	'payload' => $payload
+        ]);
+
+		$this->url = 'https://www.example.com';
+        $this->serverRequest = new \Laminas\Diactoros\ServerRequest(array(),array(), $this->url);
+
+	}
+
+	private function getWrongKey() {
+        $key = <<<EOF
```

Comment:
The wrong key could be added as a fixture, similar to the correct keys...

Reply:
I think that unit tests are often used as example code as well, so having code that generates a key is somewhat useful... but sure, I could just dump a different key in the fixtures directory. I'd rather keep it like it is however.